tests/kernel/mem_protect/userspace/src/inherit.c

```diff
@@ -0,0 +1,105 @@
+/* 
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <kernel_structs.h>
+#include <string.h>
+#include <stdlib.h>
+
+#define INHERIT_STACK_SIZE 1024
+
+/* function prototypes */
+static inline void dummy_start(struct k_timer *timer){ ARG_UNUSED(timer);}
+static inline void dummy_end(struct k_timer *timer){ ARG_UNUSED(timer);}
+
+/* Kernel objects */
+K_THREAD_STACK_DEFINE(test_1_stack, INHERIT_STACK_SIZE);
+K_SEM_DEFINE(inherit_sem, 0, 10);
+K_MUTEX_DEFINE(inhert_mutex);
+K_TIMER_DEFINE(inherti_timer, dummy_start, dummy_end);
+K_MSGQ_DEFINE(inherit_msgq, 10, 10, 2);
+__kernel struct k_thread test_1_tid;
+extern bool valid_fault;
+
+/* variables */
+#ifdef CONFIG_ARM
+u8_t __aligned(32) inherit_buf[32]; /* for mem domain */
+#else
+u8_t __aligned(4096) inherit_buf[4096]; /* for mem domain */
+#endif
+
+K_MEM_PARTITION_DEFINE(inherit_memory_partition, inherit_buf, sizeof(inherit_buf),
+                       K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *inherit_memory_partition_array[] = {
+	&inherit_memory_partition
+};
+
+struct k_mem_domain inherit_mem_domain;
+
+__attribute__((optimize("O0"))) void access_test()
+{
+	u32_t msg_q_data = 0xA5A5;
+	/* check for all accesses  */
+	k_sem_give(&inherit_sem);
+	k_mutex_lock(&inhert_mutex,K_FOREVER);
+	(void) k_timer_status_get(&inherti_timer);
+	k_msgq_put(&inherit_msgq, (void *)&msg_q_data, K_NO_WAIT);
+	k_mutex_unlock(&inhert_mutex);
+	inherit_buf[10] = 0xA5;
+}
+
+void test_thread_1_for_user(void *p1, void *p2, void *p3)
+{
+	access_test();
+	if(_is_user_context()) {
+		/* 2nd time it comes here it should raise the test case as pass. */
+		ztest_test_pass();
+	}
+	/* k_thread_abort(k_current_get()); */
+	/* exit tc gracefully */
+}
+
+void test_thread_1_for_SU(void *p1, void *p2, void *p3)
+{
+	valid_fault = false;
+	access_test();
+	//k_mem_domain_remove_thread(k_current_get());
+	/* if no protection fault has occurred then supervisor mode inherit is working*/
+	/* now to check if user mode inherit is working if control is passed from SU */
+	k_thread_user_mode_enter(test_thread_1_for_user, NULL, NULL, NULL);
+
+	/* control never comes here */
+}
+
+void inherit_test_1(void *p1, void *p2, void *p3)
+{
+	
+	k_mem_domain_init(&inherit_mem_domain, 1, inherit_memory_partition_array);
+	k_mem_domain_add_thread(&inherit_mem_domain, k_current_get());
+
+	inherit_buf[0] = 0xA5;
+		
+	k_thread_access_grant(k_current_get(),
+			      &inherit_sem,
+			      &inhert_mutex,
+			      &inherti_timer,
+			      &inherit_msgq, &test_1_stack, NULL);
+
+	k_thread_create(&test_1_tid,
+			test_1_stack,
+			INHERIT_STACK_SIZE,
+			test_thread_1_for_SU,
+			NULL, NULL, NULL,
+			0, K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_sleep(1000);
```

Comment:
We shouldn't sleep in a test case; if we need to synchronize we should do it explicitly.

Reply:
I don't think is wrong. We shouldn't use k_yield() so I am using sleep. i don't see a problem in using it this way.